testsuite/forge/src/backend/local/swarm.rs

```diff
@@ -541,3 +546,48 @@ impl Swarm for LocalSwarm {
         todo!()
     }
 }
+
+#[derive(Debug)]
+pub struct ActiveNodesGuard {
+    counter: Arc<Mutex<usize>>,
+    slots: usize,
+}
+
+impl ActiveNodesGuard {
+    pub async fn grab(slots: usize, counter: Arc<Mutex<usize>>) -> Self {
+        let max = num_cpus::get();
+        let mut idx = 0;
+        loop {
+            {
+                let mut guard = counter.lock();
+                // first check is so that if test needs more slots than cores,
+                // we still allow it to run (during low contention)
+                if *guard <= 2 || *guard + slots <= max {
+                    info!(
+                        "Grabbed {} node slots to start test, already active {} swarm nodes",
+                        slots, *guard
+                    );
+                    *guard += slots;
+                    drop(guard);
+                    return Self { counter, slots };
+                }
+                idx += 1;
+                // log only if idx is power of two, to reduce logs
+                if (idx & (idx - 1)) == 0 {
```

Comment:
We can use the `sample!` macro here, which is cleaner.

Reply:
it doesn't have exponential backoff, so I'll leave as this.

but I can add exponential backoff there too :)